### PR TITLE
feat(driver): add CacheIndexer and its Driver

### PR DESCRIPTION
**Pull request by hanxiao**
Bloom filter to test whether a doc is observed or not based on its ``doc.id``

### CODE REVIEW

**Comment by JoanFM** [comment]
Removing the document from requst may be too much? Better to flag it in case it needs to be indexed or procrssed by downstream Pods?

**Comment by hanxiao** [comment]
> Removing the document from requst may be too much? Better to flag it in case it needs to be indexed or procrssed by downstream Pods?

yes, just an example, the `hit-cache` behavior can be changed.

**Comment by hanxiao** [comment]
something like `on_hit`, `on_miss`

File: jina/drivers/cache.py [hunk 1]
```diff
@@ -1,49 +1,88 @@
...
 if False:
     from ..proto import jina_pb2
 
 
-class UniqueDocDriver(BaseRecursiveDriver):
-    """ Remove docs that previously have seen in the cache.
-
-    :class:UniqueDocDriver: follows a simple rule: it stores all new doc.id in a cache file.
-    If a doc.id hits the cache, then it is removed from the request.
+class BloomFilterDriver(BaseRecursiveDriver):
```
Comment (JoanFM) [inline jina/drivers/cache.py:15]: Is the plan to have 2 implementations? Bloom Filters can lead to false positives, thus some documents may not be indexed. This should at least be warned?

Comment (hanxiao) [reply jina/drivers/cache.py:16]: add a property to show FP rate. note, this very simple naive bloomfilter, there are more advanced bloomfilters. the general direction is correct, I will leave the extensions to others

Comment (hanxiao) [reply jina/drivers/cache.py:16]: > Is the plan to have 2 implementations? 

second thought, you are right. I will add a `set`-based checker

**Comment by hanxiao** [comment]
i'm suspending this PR for now, as it makes more sense to implement Cache as an IndexExecutor so that we can leverage existing index interface

**Pull request merged**